File: lib/src/db/entities/member.entity.ts
import { BaseEntity } from "../../common/base/base.entity";
import {
  Column,
  Entity,
  JoinColumn,
  ManyToMany,
  ManyToOne,
  PrimaryGeneratedColumn
} from "typeorm";
import { User } from "./user.entity";
import { Project } from "./project.entity";
import { IMember, MemberRole } from "@traceo/types";
import { Alert } from "./alert.entity";

@Entity()
export class Member extends BaseEntity implements IMember {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column("simple-enum", {
    enum: MemberRole,
    nullable: false
  })
  role: MemberRole;

  @ManyToOne(() => User, {
    onUpdate: "CASCADE",
    onDelete: "CASCADE"
  })
  @JoinColumn({
    name: "user_id"
  })
  user: User;

  @ManyToOne(() => Project, {
    onUpdate: "CASCADE",
    onDelete: "CASCADE"
  })
  @JoinColumn({
    name: "project_id"
  })
  project: Project;

  @ManyToMany(() => Alert)
  alerts: Alert[];
}

File: server/traceo-common/traceo-common-jpa/src/main/java/org/traceo/common/jpa/entities/MemberEntity.java
package org.traceo.common.jpa.entities;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.traceo.common.transport.enums.MemberRole;

@Entity
@Table(name = "traceo_project_member")
@Getter
@Setter
public class MemberEntity extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(nullable = false)
    private MemberRole role = MemberRole.NONE;

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "user_id")
    private UserEntity user;

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "project_id")
    private ProjectEntity project;

    public MemberEntity() {}

    public MemberEntity(UserEntity user, ProjectEntity project) {
        this(user, project, MemberRole.VIEWER);
    }

    public MemberEntity(UserEntity user, ProjectEntity project, MemberRole role) {
        this.user = user;
        this.project = project;
        this.role = role;
    }
}
